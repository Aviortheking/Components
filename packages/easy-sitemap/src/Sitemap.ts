import { ServerResponse } from 'http'

export default class Sitemap {

	private static allowedChangefreq = ['always', 'hourly', 'daily', 'weekly', 'monthly', 'yearly', 'never']

	private datas = '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">'

	public constructor(
		private domain: string, private options?: {
			response?: ServerResponse
		}
	) {
		if (this.options?.response) {
			this.options.response.setHeader('Content-Type', 'text/xml')
			this.options.response.write(this.datas)
		}
	}

	/**
	 * Add a new Entry into the Sitemap
	 * @param path the url path
	 * @param options aditional datas you want in the sitemap for the `path`
	 */
	public addEntry(path: string, options?: {
		changefreq?: 'always' | 'hourly' | 'daily' | 'weekly' | 'monthly' | 'yearly' | 'never'
		lastmod?: Date
		priority?: 1 | 0.9 | 0.8 | 0.7 | 0.6 | 0.5 | 0.4 | 0.3 | 0.2 | 0.1 | 0
		images?: Array<{
			location: string
			caption?: string
			geoLocation?: string
			title?: string
			license?: string
		}>
	}) {
		let entryString = '<url>'

		const url = this.fixText(`${this.domain}${path}`)

		entryString += `<loc>${url}</loc>`
		if (options) {
			if (options.changefreq) {
				if (!Sitemap.allowedChangefreq.includes(options.changefreq)) {
					console.warn(`changefreq is not valid (${options.changefreq})`)
				} else {
					entryString += `<changefreq>${options.changefreq}</changefreq>`
				}
			}
			if (options.lastmod) {
				entryString += `<lastmod>${this.fixText(options.lastmod.toISOString())}</lastmod>`
			}
			if (options.priority) {
				if (options.priority <= 1 && options.priority >= 0 && options.priority.toString().length <= 3) {
					entryString += `<priority>${options.priority}</priority>`
				} else {
					console.warn(`Priority is not valid (${options.priority})`)
				}
			}
			if (options.images) {
				if (options.images.length > 1000) {
					console.warn('image cant have more than 1000 images, skipping')
				} else {
					for (const image of options.images) {
						if (!image.location) {
							console.warn('Images need a Location')
							continue
						}
						entryString += '<image:image>'
						entryString += `<image:loc>${this.fixText(image.location.startsWith('/') ? `${this.domain}${image.location}` : image.location)}</image:loc>`
						entryString += this.optionalEntry('image:caption', image.caption)
						entryString += this.optionalEntry('image:geo_location', image.geoLocation)
						entryString += this.optionalEntry('image:title', image.title)
						entryString += this.optionalEntry('image:license', image.license)
						entryString += '</image:image>'
					}
				}
			}
		}
		entryString += '</url>'
		if (this.options?.response) {
			this.options.response.write(entryString)
		} else {
			this.datas += entryString
		}
	}

	/**
	 * Finish the Sitemap
	 */
	public build(): string {
		if (this.options?.response) {
			this.options.response.write('</urlset>')
			this.options.response.end()
			return ''
		}
		return this.datas + '</urlset>'
	}

	private optionalEntry(tag: string, entry?: string) {
		return entry ? `<${tag}>${this.fixText(entry)}</${tag}>` : ''
	}

	private fixText(txt: string): string {
		return txt.replace(/&/g, '&amp;')
			.replace(/'/g, '&apos;')
			.replace(/"/g, '&quot;')
			.replace(/>/g, '&gt;')
			.replace(/</g, '&lt;')
	}
}
