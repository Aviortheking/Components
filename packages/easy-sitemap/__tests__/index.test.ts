/// <reference types="jest" />

import Sitemap from '../src/Sitemap'

describe('Basic Sitemap Tests', () => {
	it('should return an empty sitemap', () => {
		const sitemap = new Sitemap('https://www.example.com')
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should return a basic sitemap', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path')
		sitemap.addEntry('/')
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should return a sitemap', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			changefreq: 'always',
			lastmod: new Date('2021-01-20'),
			priority: 1
		})
		sitemap.addEntry('/')
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should not add priority when it is incorrect', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			// @ts-expect-error
			priority: 255
		})
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should not add changefreq if value is incorrect', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			// @ts-expect-error
			changefreq: 'pouet'
		})
		expect(sitemap.build()).toMatchSnapshot()
	})
})

describe('image:image tests', () => {
	it('should build corretcly with single image:image', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			images: [{
				location: '/test',
				geoLocation: 'Nantes',
				title: 'Title',
				caption: 'Image Caption',
				license: 'Example license url'
			}]
		})
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should skip image:image if no location is set', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			// @ts-expect-error
			images: [{
				geoLocation: 'Nantes',
			}]
		})
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should skip image:image if there is more than 1000 images', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			images: Array.from(new Array(1001)).map(() => ({
				location: '/test',
			}))
		})
		expect(sitemap.build()).toMatchSnapshot()
	})
	it('should build corretcly with multiple image:image', () => {
		const sitemap = new Sitemap('https://www.example.com')
		sitemap.addEntry('/path', {
			images: [{
				location: '/test',
				geoLocation: 'Nantes',
				title: 'Title',
				caption: 'Image Caption',
				license: 'Example license url'
			}, {
				location: '/test-2',
				geoLocation: 'Paris',
				title: 'Title2',
				caption: 'Image Caption2',
				license: 'Example license url'
			}]
		})
		expect(sitemap.build()).toMatchSnapshot()
	})
})
