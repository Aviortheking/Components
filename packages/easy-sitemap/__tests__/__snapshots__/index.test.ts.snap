// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Sitemap Tests should not add changefreq if value is incorrect 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc></url></urlset>"`;

exports[`Basic Sitemap Tests should not add priority when it is incorrect 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc></url></urlset>"`;

exports[`Basic Sitemap Tests should return a basic sitemap 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc></url><url><loc>https://www.example.com/</loc></url></urlset>"`;

exports[`Basic Sitemap Tests should return a sitemap 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc><changefreq>always</changefreq><lastmod>2021-01-20T00:00:00.000Z</lastmod><priority>1</priority></url><url><loc>https://www.example.com/</loc></url></urlset>"`;

exports[`Basic Sitemap Tests should return an empty sitemap 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"></urlset>"`;

exports[`image:image tests should build corretcly with multiple image:image 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc><image:image><image:loc>https://www.example.com/test</image:loc><image:caption>Image Caption</image:caption><image:geo_location>Nantes</image:geo_location><image:title>Title</image:title><image:license>Example license url</image:license></image:image><image:image><image:loc>https://www.example.com/test-2</image:loc><image:caption>Image Caption2</image:caption><image:geo_location>Paris</image:geo_location><image:title>Title2</image:title><image:license>Example license url</image:license></image:image></url></urlset>"`;

exports[`image:image tests should build corretcly with single image:image 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc><image:image><image:loc>https://www.example.com/test</image:loc><image:caption>Image Caption</image:caption><image:geo_location>Nantes</image:geo_location><image:title>Title</image:title><image:license>Example license url</image:license></image:image></url></urlset>"`;

exports[`image:image tests should skip image:image if no location is set 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc></url></urlset>"`;

exports[`image:image tests should skip image:image if there is more than 1000 images 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" image:xmlns=\\"http://www.google.com/schemas/sitemap-image/1.1\\"><url><loc>https://www.example.com/path</loc></url></urlset>"`;
